posts/2019-02-20-framework-release-v138-websockets.md

```diff
@@ -0,0 +1,161 @@
+
+```yml
+functions:
+  default:
+    handler: handler.default
+    events:
+      - websocket: $default
+```
+
+This would create a `$default` route that would forward all websocket events (including `$connect` and `$disconnect`) to your `default` function.
+
+You can also specify your event as an object and add more routes:
+
+
+```yml
+functions:
+  connect:
+    handler: handler.connect
+    events:
+        - websocket:
+            route: $connect
+  disconnect:
+    handler: handler.disconnect
+    events:
+      - websocket:
+          route: $disconnect
+  default:
+    handler: handler.default
+    events:
+      - websocket:
+          route: $default
+  echo:
+    handler: handler.echo
+    events:
+      - websocket:
+          route: $echo
+
+```
+
+The object notation will be useful when you want to add more configuration to your websocket events, for example when we add support for authorizers in an upcoming release.
+
+Once you deploy this service, you'll see the endpoint of your websocket backend in your terminal. Using this endpoint, you can connect to your websocket backend using any websocket client.
+
+By default, the route selection expression is set to `$request.body.action`. This property tells API Gateway how to parse the data coming into your websocket endpoint. So with this default behavior, and using the service above, you can invoke the `echo` function by using the following JSON object as your websocket event body: 
+
+```json
+{
+  "action": "echo",
+  "data":  "hello"
+}
+```
+ 
+You can overwrite the route expression by specifying the `websocketsApiRouteSelectionExpression` key in the `provider` object:
+
+```yml
+provider:
+  name: aws
+  runtime: nodejs8.10
+  websocketsApiRouteSelectionExpression: "$request.body.route"
+```
+
+In that case, your websocket body should be:
+
+```json
+{
+  "route": "echo",
+  "data":  "hello"
+}
+```
+
+Remember that any other body/data coming to your websocket back would invoke the default function.
+
+The framework also takes care of setting the permissions required for your lambda function to communicate to the connected clients. So you'll be able to send data to any client right away by using the new `ApiGatewayManagementApi` Service, without having to worry about IAM policies.
+
+```js
+  const client = new AWS.ApiGatewayManagementApi({
+    apiVersion: '2018-11-29',
+    endpoint: `https://${event.requestContext.domainName}/${event.requestContext.stage}`
+  });
+
+  await client
+    .postToConnection({
+      ConnectionId: event.requestContext.connectionId,
+      Data: `echo route received: ${event.body}`
+    })
+    .promise();
+```
+
+**Note:** At the time of this writing, the Lambda runtime does not include the latest version of the AWS SDK that includes this new `ApiGatewayManagementApi` service. So you'll have to deploy your own by adding it to your `package.json`.
+
```

Comment:
Might be wortwhile to point out that `http` and `websocket` can live side-by-side in a single service without any issues since multiple Gateways will be created in that case. This was something I was not aware of...

Reply:
good point.